SEP-0012.md

```diff
@@ -0,0 +1,495 @@
+This could be the most common FITS WCS standard, or gWCS or custom implementations for specific facilites with different requirements.
+
+Having made this change, which was a large breaking change from the original API, we reviewed all the existing functionality and API as it was adapted to use the high level WCS interface and the result is presented below.
+
+# Detailed Description
+
+The API presented here is limited to the two core classes `NDCubeABC` and `ExtraCoordsABC`.
+This is primarily an attempt to limit the discussion in this SEP to the core API and not implementation decisions taken in the development of ndcube.
+If people are interested in the implementation of the proposed API large amounts of it are implemented on the master branch of the [ndcube repository](https://github.com/sunpy/ndcube) or open pull requests.
+
+The `NDCube` class inherits the astropy [`NDDataBase`](https://github.com/astropy/astropy/blob/master/astropy/nddata/nddata_base.py#L11) class, including the requirement that the `.wcs` property be a `BaseHighLevelWCS` instance.
+In the following code snippets of the API use the Python type hint syntax as appropriate to highlight accepted types and return types, these are not enforced at runtime.
+
+
+## NDCube
+
+Below is a class definition for the Abstract Base Class (ABC) for `NDCube`.
+Its purpose is to specify the API as completely as possible, including input and return types. 
+The behaviour of all the methods should be described by their included docstrings, a version of this (likely out of sync) has been [PRed to ndcube](https://github.com/sunpy/ndcube/pull/304).
+
+```python
+
+class NDCubeABC(astropy.nddata.NDDataBase):
+
+    @abc.abstractmethod
+    def __init__(self,
+                 data: "numpy.typing.ArrayLike",
+                 uncertainty: Any = None,
+                 mask: Any = None,
+                 wcs: Union[BaseLowLevelWCS, BaseHighLevelWCS] = None,
+                 meta: Mapping = None,
+                 unit: Union[str, u.Unit] = None,
+                 copy: bool = False,
+                 extra_coords: "ExtraCoordsABC" = None):
+        pass
+
+    @abc.abstractmethod
+    def __getitem__(self, item: Union[int, slice, Iterable[Union[slice, int]]]) -> "NDCubeABC":
+        """
+        An ndcube object must be sliceable in a similar way to a numpy array,
+        however, support for a `step` is not required and no "fancy indexing" is allowed.
+
+        When a slice is applied to a NDCube object it must be applied to the following attributes:
+
+          * data
+          * wcs
+          * uncertainty
+          * mask
+          * extra_coords
+
+        the following attributes must be copied unaltered to the returned NDCube:
+
+          * unit
+          * meta
+
+        """
+
+    @abc.abstractmethod
+    @property
+    def extra_coords(self) -> ExtraCoordsABC:
```

Comment:
Exactly what `extra_coords` are should be explained at a high level before this section.

Reply:
There is always going to be a race condition between parts of this doc, these properties exist and their return values are described in more detail later. If it were to be flipped then the issue would present in reverse, you would have no context for the other classes.

I can add some more context to the top of the document though.

Reply:
I agree with @ehsteve, and there currently isn't any high level introduction to extra coordinates at the top.